transform_field/__init__.py

```diff
@@ -207,6 +224,77 @@ def consume(self, reader):
             self.handle_line(line)
         self.flush()
 
+    def validate(self, catalog: Catalog):
+        """
+        Validate the transformations by checking if each transformation type is compatible with the column type
+        :param catalog: the catalog of streams with their json schema
+        """
+        LOGGER.info('Starting validation of transformations...')
+
+        if not catalog:
+            raise CatalogRequiredException('Catalog missing! please provide catalog to run validation.')
+
+        # get the schema of each stream
+        schemas = utils.get_stream_schemas(catalog)
+
+        for stream_id in self.trans_meta:
+            self.__validate_stream_trans(stream_id, schemas.get(stream_id))
+
+    def __validate_stream_trans(self, stream_id: str, stream_schema: Union[Schema, Dict]):
+        """
+        Validation of each stream's transformations
+        :param stream_id: ID of the stream
+        :param stream_schema: schema of the streams
+        """
+
+        if stream_id not in self.trans_meta:
+            return
+
+        # check if we even have schema for stream of this transformation
+        if stream_schema is None:
+            raise StreamNotFoundException(stream_id)
+
+        # check if we stream has not empty schema
+        if not stream_schema:
+            raise NoStreamSchemaException(stream_id)
+
+        for transformation in self.trans_meta[stream_id]:
+            trans_type = transformation.type
+            field_id = transformation.field_id
+
+            if isinstance(stream_schema, Schema):
+                field_type = stream_schema.properties[field_id].type
+                field_format = stream_schema.properties[field_id].format
+            else:
+                field_type = stream_schema['properties'][field_id].get('type')
+                field_format = stream_schema['properties'][field_id].get('format')
+
+            if trans_type in (TransformationTypes.HASH.value, TransformationTypes.MASK_HIDDEN.value) or \
+                    trans_type.startswith(TransformationTypes.HASH_SKIP_FIRST.value):
+                if not (field_type is not None and 'string' in field_type and not field_format):
+                    raise InvalidTransformationException(
+                        f'Cannot apply `{trans_type}` transformation type to a non-string field `'
+                        f'{field_id}` in stream `{stream_id}`')
```

Comment:
should we worry about length of strings?

Taps put max string length into catalog. tap-mysql [here](https://github.com/transferwise/pipelinewise-tap-mysql/blob/5a9e043b2581d01229970f8b9debf3452a877515/tap_mysql/discover_utils.py#L208), tap-postgres [here](https://github.com/transferwise/pipelinewise-tap-postgres/blob/18406d5be0e4dbef397b54078a55af9db9184917/tap_postgres/discovery_utils.py#L269).

For example hashing a 3 characters length string produces a longer hashdigest than 3 characters. The output string breaks the definition in the catalog. However it's fine for the current target implementations because they're generating long enough varchar commands and JSONschema record validation is turned off.

Wdyt, should we deal with this scenario here, or fine to leave this to the target implemention?

Reply:
Thanks.

It is indeed a concern too, however, as you said, the targets don't currently enforce it so we can deal with it when it's an actual problem 😅 .